src/main/java/bio/terra/janitor/db/JanitorDao.java

```diff
@@ -69,16 +69,110 @@ public TrackedResourceId createResource(
+   * TrackedResourceState#CLEANING}. Inserts a new {@link CleanupFlight} for that resource as
+   * initiating.
+   */
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public Optional<TrackedResource> updateResourceForCleaning(Instant now, String flightId) {
+    UUID id =
+        DataAccessUtils.singleResult(
+            jdbcTemplate.queryForList(
+                "SELECT id FROM tracked_resource "
+                    + "WHERE state = :ready and expiration <= :now LIMIT 1",
+                new MapSqlParameterSource()
+                    .addValue("ready", TrackedResourceState.READY.toString())
+                    .addValue("now", now.atOffset(ZoneOffset.UTC)),
+                UUID.class));
+    if (id == null) {
+      // No resources ready to schedule.
+      return Optional.empty();
+    }
+
+    TrackedResourceId resourceToClean = TrackedResourceId.create(id);
+    jdbcTemplate.update(
+        "INSERT INTO cleanup_flight (tracked_resource_id, flight_id, flight_state) "
+            + "VALUES (:tracked_resource_id, :flight_id, :initiating)",
+        new MapSqlParameterSource()
+            .addValue("tracked_resource_id", resourceToClean.uuid())
+            .addValue("flight_id", flightId)
+            .addValue("initiating", CleanupFlightState.INITIATING.toString()));
+    return Optional.of(
+        jdbcTemplate.queryForObject(
+            "UPDATE tracked_resource SET state = :cleaning WHERE id = :id "
+                + "RETURNING id, resource_uid, creation, expiration, state",
+            new MapSqlParameterSource()
+                .addValue("cleaning", TrackedResourceState.CLEANING.toString())
+                .addValue("id", resourceToClean.uuid()),
+            TRACKED_RESOURCE_ROW_MAPPER));
+  }
+
+  /** Returns up to {@code limit} resources with a cleanup flight in the given state. */
+  public List<TrackedResourceAndFlight> retrieveResourcesWith(
+      CleanupFlightState flightState, int limit) {
+    String sql =
+        "SELECT tr.id, tr.resource_uid, tr.creation, tr.expiration, tr.state, "
+            + "cf.flight_id, cf.flight_state FROM tracked_resource tr "
+            + "JOIN cleanup_flight cf ON tr.id = cf.tracked_resource_id "
+            + "WHERE cf.flight_state = :flight_state LIMIT :limit";
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("flight_state", flightState.toString())
+            .addValue("limit", limit);
+    return jdbcTemplate.query(
+        sql,
+        params,
+        (rs, rowNum) ->
+            TrackedResourceAndFlight.create(
+                TRACKED_RESOURCE_ROW_MAPPER.mapRow(rs, rowNum),
+                CLEANUP_FLIGHT_ROW_MAPPER.mapRow(rs, rowNum)));
+  }
+
+  /** Retrieve the {@link CleanupFlight}s associated with the tracked resource. */
+  public List<CleanupFlight> getFlights(TrackedResourceId trackedResourceId) {
```

Comment:
getFlightsByResourceId? Just in case we need other methods to query ClentupFlight by other things?

Reply:
I think `ByResourceId` encoded in the input parameters. I'm ok overloading this method, or we can differentiate later when we add more methods.